### PR TITLE
[Doc] add flat json doc - 3.3

### PR DESCRIPTION
**Pull request by Seaven**
Why I'm doing:

What I'm doing:
- add 3.3 FlatJson.md and sidebar
- add get_json_bool function and update function list
- add related session variable and BE para

Fixes #issue

## What type of PR is this:

- [ ] BugFix
- [ ] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [x] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [ ] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function
- [ ] This is a backport pr

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [x] 3.3
  - [ ] 3.2
  - [ ] 3.1
  - [ ] 3.0
  - [ ] 2.5

### CODE REVIEW

**Comment by Seaven** [comment]
> this may be a related PR and description: #31750 mechanism for removing a BE from the blacklist

the related PR:
FlatJson: https://github.com/StarRocks/starrocks/pull/37577
Blacklist: https://github.com/StarRocks/starrocks/pull/39064

File: docs/zh/using_starrocks/Flat_json.md [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+StarRocks 当前对Flat JSON的支持：
+* 导入时，支持提取公共字段，单独存储为 JSON 类型，未实现类型推导。
+* 目前只支持最顶层 JSON 字段的提取。
+* 目前会同时存储提取列，和原始 JSON 数据。
+* 兼容历史数据，无须重新导入。
+* 历史表导入新数据时，自动通过 Compaction 完成 Flat JSON操作。
+
+使用 Flat Json 需要在 BE 配置：
+```plain text
+enable_flat_json = true
```
Comment (wangruin) [inline docs/zh/using_starrocks/Flat_json.md:42]: 是 enable_json_flat ？

File: docs/zh/using_starrocks/Flat_json.md [hunk 2]
```diff
@@ -0,0 +1,80 @@
+本文介绍 Flat JSON 的基础概念，以及如何使用。
+
+自 2.2.0 版本起，StarRocks 支持 JSON 数据类型，用于支持更加灵活的数据存储。 使用JSON 类型的原因大多数是因为需要存储数据的结构易变，难以维护结构变更，但在查询 JSON 时，大部分场景并不是直接读取完整的 JSON 数据，而是访问指定的路径下数据， 例如：
+```
+// 将日志中必填的字段存储为固定字段，其他经常随业务变更的字段打包为Json存储
+select time, evevnt, user, get_json_string(remain_json, "$.from_system"), get_json_string(remain_json, "$.tag")
+from xxx_logs
+```
+
+但由于 JSON 类型的特殊性，但在查询中 JSON 类型的性能表现并不如标准类型（INT，STRING等）快，其原因有：
```
Comment (wangruin) [inline docs/zh/using_starrocks/Flat_json.md:10]: 但是由于 JSON 类型的特殊性，在....  怎么样？

File: docs/zh/using_starrocks/Flat_json.md [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+* JSON 类型是半结构化类型，在存储上需要存储每行数据的结构信息，存储占用多，压缩效率低。
+* 在查询时，需要根据运行时数据检测使用结构信息，难以实现向量化执行优化。
+* 在查询时，需要读取完整的 Json 数据，其中包含了大量冗余的字段。
+
+所以 StarRocks 从3.3起，支持Flat Json功能，旨在优化 JSON 的查询性能。
+
+## 什么是 Flat Json
+
+Flat JSON 的核心原理，是在导入时检测 JSON 数据，将 JSON 数据中公共的字段提取为标准类型数据存储，在查询JSON时，通过这些公共字段数据优化 JSON 的查询速度，例如：
```
Comment (wangruin) [inline docs/zh/using_starrocks/Flat_json.md:19]: `核心原理，是在`  这里没必要 逗号？
查询JSON使，这里前后 少空格

File: docs/zh/using_starrocks/Flat_json.md [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+```plain text
+1, {"a": 1, "b": 21, "c": 3, "d": 4}
+2, {"a": 2, "b": 22, "d": 4}
+3, {"a": 3, "b": 23, "d": [1, 2, 3, 4]}
+4, {"a": 4, "b": 24, "d": null}
+5, {"a": 5, "b": 25, "d": null}
+6, {"c": 6, "d": 1}
+```
+
+在导入上述这样一组 JSON 数据时，`a` 和 `b`两个字段在大部分的 JSON 数据中都存在并且其数据类型相似（都是int），那么就可以将`a`，`b`两个字段的数据都从 JSON 中读取出来，单独存储为两列INT。当查询中使用到这两列时，就可以直接读取`a`,`b`两列的数据，减少查询时读取 JSON 中额外的字段数据，在计算时减少对 JSON 结构的处理开销。
```
Comment (wangruin) [inline docs/zh/using_starrocks/Flat_json.md:29]: INT 前后空格

File: docs/zh/using_starrocks/Flat_json.md [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+
+## 使用 Flat Json
+
+StarRocks 当前对Flat JSON的支持：
+* 导入时，支持提取公共字段，单独存储为 JSON 类型，未实现类型推导。
+* 目前只支持最顶层 JSON 字段的提取。
+* 目前会同时存储提取列，和原始 JSON 数据。
```
Comment (wangruin) [inline docs/zh/using_starrocks/Flat_json.md:36]: 这里也没必要加 逗号？

File: docs/zh/using_starrocks/Flat_json.md [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+* 兼容历史数据，无须重新导入。
+* 历史表导入新数据时，自动通过 Compaction 完成 Flat JSON操作。
+
+使用 Flat Json 需要在 BE 配置：
```
Comment (wangruin) [inline docs/zh/using_starrocks/Flat_json.md:40]: 加上  配置以下字段，然后重启 BE 生效？

File: docs/zh/using_starrocks/Flat_json.md [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+```plain text
+enable_flat_json = true
+```
+
+开启后新导入的 JSON 数据会自动打平。
+
+在查询 SQL 时，设置Session变量：
```
Comment (wangruin) [inline docs/zh/using_starrocks/Flat_json.md:47]: Session 前后空格

File: docs/zh/using_starrocks/Flat_json.md [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+``` sql
+set cbo_prune_subfield = true;
```
Comment (wangruin) [inline docs/zh/using_starrocks/Flat_json.md:49]: cbo_prune_subfield 本身不可见，而且我看 3.2 默认为true了，flat json 3.3 支持，这里是不是可以不提。一个功能要2个开关控制，也不易用

File: docs/zh/using_starrocks/Flat_json.md [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+
+其中几个相关指标：
+* `PushdownAccessPaths`：下推存储的子字段路径数量。
+* `AccessPathHits`：命中 Flat JSON 优化的存储文件数量，其子项详细打印了具体命中的 JSON。 
+* `AccessPathUnhits`：未命中 Flat JSON 优化的存储文件数据量，其子项详细打印了具体未命中的 JSON 。
+* `JsonFlattern`：当存在未命中 Flat JSON 优化时，执行动态 Flat JSON 的耗时。
+
+## 其他配置
+
+BE 上：
```
Comment (wangruin) [inline docs/zh/using_starrocks/Flat_json.md:67]: 这里这么多BE 配置命名为什么是 json_flat，而不是 flat_json

File: docs/zh/using_starrocks/Flat_json.md [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+* `json_flat_column_max`：控制 Flat JSON 时，最多提取的子列数量，默认为 20。
+* `json_flat_sparsity_factor`：控制 Flat JSON 时，同名列的占比阈值，当同名列占比低于该值不进行提取，默认为0.9。
+
+## 注意事项
+
+* 开启 Flat JSON 后，提取列会占用额外的存储资源。
+* 开启 Flat JSON 后，会加大导入 JSON 的耗时，提取的 JSON 越多，耗时越高。
+* 开启 Flat JSON 后，Compaction 的耗时和内存使用量会增高。
+* `cbo_prune_json_subfield` 只有在命中 Flat JSON 时才有效果，其他情况下可能存在性能负优化。
+* 如果遇到 Crash 或者查询报错的问题，可以通过关闭该功能 以及 Session 变量规避。
```
Comment (wangruin) [inline docs/zh/using_starrocks/Flat_json.md:80]: 通过 Session 变量关闭功能来规避？

File: docs/zh/using_starrocks/Flat_json.md [hunk 11]
```diff
@@ -0,0 +1,80 @@
+本文介绍 Flat JSON 的基础概念，以及如何使用。
+
+自 2.2.0 版本起，StarRocks 支持 JSON 数据类型，用于支持更加灵活的数据存储。 使用JSON 类型的原因大多数是因为需要存储数据的结构易变，难以维护结构变更，但在查询 JSON 时，大部分场景并不是直接读取完整的 JSON 数据，而是访问指定的路径下数据， 例如：
```
Comment (wangruin) [inline docs/zh/using_starrocks/Flat_json.md:3]: 使用JSON 类型的原因  空格

File: docs/zh/using_starrocks/Flat_json.md [hunk 12]
```diff
@@ -0,0 +1,80 @@
...
+* `PushdownAccessPaths`：下推存储的子字段路径数量。
+* `AccessPathHits`：命中 Flat JSON 优化的存储文件数量，其子项详细打印了具体命中的 JSON。 
+* `AccessPathUnhits`：未命中 Flat JSON 优化的存储文件数据量，其子项详细打印了具体未命中的 JSON 。
+* `JsonFlattern`：当存在未命中 Flat JSON 优化时，执行动态 Flat JSON 的耗时。
+
+## 其他配置
+
+BE 上：
+* `enable_json_flat`：控制 Flat JSON 是否开启，默认为 false。
+* `json_flat_null_factor`：控制 Flat JSON 时，提取列的 NULL 值占比阈值，高于该比例不进行提取，默认为0.3。
```
Comment (wangruin) [inline docs/zh/using_starrocks/Flat_json.md:69]: 0.3 前面空格

File: docs/zh/using_starrocks/Flat_json.md [hunk 13]
```diff
@@ -0,0 +1,80 @@
...
+4, {"a": 4, "b": 24, "d": null}
+5, {"a": 5, "b": 25, "d": null}
+6, {"c": 6, "d": 1}
+```
+
+在导入上述这样一组 JSON 数据时，`a` 和 `b`两个字段在大部分的 JSON 数据中都存在并且其数据类型相似（都是int），那么就可以将`a`，`b`两个字段的数据都从 JSON 中读取出来，单独存储为两列INT。当查询中使用到这两列时，就可以直接读取`a`,`b`两列的数据，减少查询时读取 JSON 中额外的字段数据，在计算时减少对 JSON 结构的处理开销。
+
+## 使用 Flat Json
+
+StarRocks 当前对Flat JSON的支持：
```
Comment (wangruin) [inline docs/zh/using_starrocks/Flat_json.md:33]: Flat JSON 前后空格

File: docs/zh/using_starrocks/Flat_json.md (hunk 1 above)
Comment (Seaven) [reply docs/zh/using_starrocks/Flat_json.md:42]: yeah

File: docs/zh/using_starrocks/Flat_json.md (hunk 8 above)
Comment (Seaven) [reply docs/zh/using_starrocks/Flat_json.md:49]: 可以

**Comment by EsoragotoSpirit** [comment]
BE BLACKLIST added via https://github.com/StarRocks/starrocks/pull/44884

**Review by evelynzhaojie — DISMISSED** [review]

File: docs/en/administration/management/BE_configuration.md [hunk 14]
```diff
@@ -3204,6 +3204,51 @@ When this value is set to less than `0`, the system uses the product of its abso
...
+- Description: Whether to enable the Flat JSON feature. After this feature is enabled, newly loaded JSON data will be automatically flattened, improving JSON query performance.
+- Introduced in: v3.3.0
+
+##### json_flat_null_factor
+
+- Default: 0.3
+- Type: Double
+- Unit:
+- Is mutable: Yes
+- Description: The proportion of NULL values in the column to extract. A column will not be extracted if its proportion of NULL value is higher than this threshold.
```
Comment (EsoragotoSpirit) [inline docs/en/administration/management/BE_configuration.md:3222]: 这个参数的解释看不出是 Flat JSON 相关的？另外这几个参数需要加一下依赖关系吗？takes effect when enable_json_flat is set
 to ture

**Review by evelynzhaojie — DISMISSED** [review]

File: docs/en/using_starrocks/Flat_json.md [hunk 15]
```diff
@@ -0,0 +1,185 @@
...
+
+The `a` and `b` fields exist in most rows and their data types are similar (both INT). Therefore, the data of the `a` and `b` fields can be read from the JSON data and stored separately as two INT columns. If a query is destined for these two columns, the data can be directly read without the need to read other columns, reducing the processing overhead of JSON structure during computation.
+
+## Usage notes
+
+- Currently, only StarRocks shared-nothing clusters support Flat JSON. All StarRocks table types support Flat JSON.
+- During data loading, this feature extracts common fields and stores them separately as JSON data. However, type inference is not supported. StarRocks will continuously iterate the Flat JSON feature to support type inference and storage optimization in future versions.
+- Currently, only the top-level JSON fields can be extracted.
+- Both the extracted columns and the original JSON data will be stored. The extracted data is removed when the original data is deleted.
+- Historical data that have been loaded before will not be loaded again.
```
Comment (EsoragotoSpirit) [inline docs/en/using_starrocks/Flat_json.md:50]: will not be loaded again 不太能看出来兼容历史数据的意思。

Historical data will be automatically flattened/extracted? 不确定我的理解是否正确

**Review by EsoragotoSpirit — DISMISSED** [review]

**Review by EsoragotoSpirit — APPROVED** [review]

**Review by evelynzhaojie — APPROVED** [review]

**Pull request merged**